llvm/test/CodeGen/AIE/aie2/internalize.ll

```diff
@@ -0,0 +1,123 @@
+define void @func_used_alwaysinline(ptr addrspace(1) %out, i32 %val) #1 {
+entry:
+  store volatile i32 %val, ptr addrspace(1) %out
+  ret void
+}
+
+; ONLYMAIN: declare void @decl_func()
+; FUNCLIST-NOT: declare void @decl_func()
+declare void @decl_func()
+
+; ONLYMAIN: define{{.*}}void @main()
+; ONLYMAIN: tail call fastcc void @func_used_noinline
+; ONLYMAIN: store volatile
+; ONLYMAIN: tail call fastcc void @decl_func()
+; ONLYMAIN: ret void
+; FUNCLIST-NOT: define{{.*}}void @main()
+; FUNCLIST-NOT: tail call fastcc void @func_used_noinline
+; FUNCLIST-NOT: store volatile
+; FUNCLIST-NOT: tail call fastcc void @decl_func()
+; FUNCLIST-NOT: ret void
+define void @main() {
+entry:
+    %val = add i32 42, 0
+    tail call fastcc void @func_used_noinline(ptr addrspace(1) @gvar_used, i32 %val)
+    tail call fastcc void @func_used_alwaysinline(ptr addrspace(1) @gvar_used, i32 %val)
+    tail call fastcc void @decl_func()
+    ret void
+}
+
+; ONLYMAIN-NOT: define{{.*}}void @func_used_skipfunc1
+; FUNCLIST: define{{.*}}void @func_used_skipfunc1
+define void @func_used_skipfunc1(ptr addrspace(1) %out, i32 %val) #0 {
+entry:
+  store i32 %val, ptr addrspace(1) %out
+  ret void
+}
+
+; ONLYMAIN-NOT: define{{.*}}void @skipfunc1()
+; ONLYMAIN-NOT: tail call fastcc void @func_used_skipfunc1
+; ONLYMAIN-NOT: ret void
+; FUNCLIST: define{{.*}}void @skipfunc1()
+; FUNCLIST: tail call fastcc void @func_used_skipfunc1
+; FUNCLIST: ret void
+define void @skipfunc1() {
+entry:
+    %val = add i32 42, 0
+    tail call fastcc void @func_used_skipfunc1(ptr addrspace(1) @gvar_used, i32 %val)
+    ret void
+}
+
+; ONLYMAIN-NOT: define{{.*}}void @func_used_skipfunc2
+; FUNCLIST: define{{.*}}void @func_used_skipfunc2
+define void @func_used_skipfunc2(ptr addrspace(1) %out, i32 %val) #0 {
+entry:
+  store i32 %val, ptr addrspace(1) %out
+  ret void
+}
+
+; ONLYMAIN-NOT: define{{.*}}void @skipfunc2()
+; ONLYMAIN-NOT: tail call fastcc void @func_used_skipfunc2
+; ONLYMAIN-NOT: ret void
+; FUNCLIST: define{{.*}}void @skipfunc2()
+; FUNCLIST: tail call fastcc void @func_used_skipfunc2
+; FUNCLIST: ret void
+define void @skipfunc2() {
+entry:
+    %val = add i32 42, 0
+    tail call fastcc void @func_used_skipfunc2(ptr addrspace(1) @gvar_used, i32 %val)
+    ret void
+}
+
+; ONLYMAIN-NOT: define{{.*}}void @func_used_skipfunc3
+; FUNCLIST: define{{.*}}void @func_used_skipfunc3
+define void @func_used_skipfunc3(ptr addrspace(1) %out, i32 %val) #0 {
+entry:
+  store i32 %val, ptr addrspace(1) %out
+  ret void
+}
+
+; ONLYMAIN-NOT: define{{.*}}void @skipfunc3()
+; ONLYMAIN-NOT: tail call fastcc void @func_used_skipfunc3
+; ONLYMAIN-NOT: ret void
+; FUNCLIST: define{{.*}}void @skipfunc3()
+; FUNCLIST: tail call fastcc void @func_used_skipfunc3
+; FUNCLIST: ret void
+define void @skipfunc3() {
+entry:
+    %val = add i32 42, 0
+    tail call fastcc void @func_used_skipfunc3(ptr addrspace(1) @gvar_used, i32 %val)
+    ret void
+}
```

Comment:
Maybe add a test where the address of a function escapes (e.g. it gets stored somewhere)?